app/actions/contactsActions.js

```diff
@@ -5,13 +5,93 @@ import { has, isEmpty, keys, values, indexOf, zipObject, omit } from 'lodash-es'
 
 import { getStorage, setStorage } from '../core/storage'
 
+export const ID = 'contacts'
+
+export type TChains = { contactKey: string, chain: string }[]
 type Contacts = {
   [name: string]: string,
 }
 
+const STORAGE_KEY_CHAIN = 'chainBook'
 const STORAGE_KEY = 'addressBook'
 
-const getContacts = async (): Promise<Contacts> => getStorage(STORAGE_KEY)
+const setChainInContactsLessChain = async () => {
+  const contacts: Contacts = await getStorage(STORAGE_KEY)
+  const chains: TChains = await getStorage(STORAGE_KEY_CHAIN)
+  const contactChains = Array.isArray(chains) ? chains : []
+  await Promise.all(
+    Object.entries(contacts).map(async ([name]) => {
+      if (!contactChains.find(chain => chain.contactKey === name)) {
+        contactChains.push({ contactKey: name, chain: 'neo2' })
+        await setStorage(STORAGE_KEY_CHAIN, contactChains)
+      }
+    }),
+  )
+}
+
+export const getChains = async (): Promise<TChains> => {
+  await setChainInContactsLessChain()
+  const chains = await getStorage(STORAGE_KEY_CHAIN)
+  return Array.isArray(chains) ? chains : []
+}
+
+const setChainContact = async (contactKey: string, chain: string) => {
+  const chains = await getChains()
+  const chainContacts: TChains = Array.isArray(chains) ? chains : []
+  if (contactKey && chain) {
+    chainContacts.push({ contactKey, chain })
+    await setStorage(STORAGE_KEY_CHAIN, chainContacts)
+  }
+}
+
+const updateChainContact = async (
+  contactKey: string,
+  chain: string,
+  oldContactKey: string,
+) => {
+  const chains = await getChains()
+  const chainContacts: TChains = Array.isArray(chains) ? chains : []
+  if (contactKey && chain && oldContactKey) {
+    const chainContactFound = chainContacts.find(
+      chainContact => chainContact.contactKey === oldContactKey,
+    )
+    if (chainContactFound) {
+      chainContacts[
+        chainContacts.indexOf(chainContactFound)
+      ].contactKey = contactKey
+      chainContacts[chainContacts.indexOf(chainContactFound)].chain = chain
+
+      await setStorage(STORAGE_KEY_CHAIN, chainContacts)
+    } else {
+      await setChainContact(contactKey, chain)
+    }
+  }
+}
+
+const deleteChainContact = async (contactKey: string) => {
+  const chains = await getChains()
+  const chainContacts: TChains = Array.isArray(chains) ? chains : []
+  if (contactKey) {
+    const newChainContacts = chainContacts.filter(
+      chainContact => chainContact.contactKey !== contactKey,
+    )
+
+    await setStorage(STORAGE_KEY_CHAIN, newChainContacts)
+  }
+}
+
+export const getContactChainAction = createActions(
+  ID,
+  (): Promise<TChains> => getChains(),
+)
+
+export const cleanContacts = async () => {
+  await setStorage(STORAGE_KEY, {})
+  await setStorage(STORAGE_KEY_CHAIN, [])
+}
+
+export const getContacts = async (): Promise<Contacts> =>
+  getStorage(STORAGE_KEY)
```

Comment:
without chain data this is always going to return the neo legacy contacts 👆

Reply:
about it function, you prefer to return already filtered by chain contacts, is that it? because in contact screen I make the filter and show only contacts that own a specific chain

Reply:
We will want the redux state tree to always have the right contacts in place so we need to filter one level higher, I am afraid filtering at the component level will lead to more work and unexpected bugs

Reply:
ok, I understood, so I gonna the filter in high level